### ISSUE TITLE
org-mode integration

### ISSUE THREAD

**Issue by fancycade** [felko/neuron-mode#26]
Hi there!

Thanks for all of the effort on this project so far. This is the only emacs zettelkasten I've got to work out of the box.

One feature that I think might be missing, or maybe it is simply a matter of explaining in the docs, is how to use org-mode with neuron-mode.

The reason I bring this up is that of course emacs users tend to love org-mode, and this was one attractive quality of this project as well:
https://www.eliasstorms.net/zetteldeft/

**Comment by srid** [comment]
The bulk of the work for this needs to be done in neuron. neuron `master` now uses Pandoc AST, so it shouldn't be too hard. The markdown is parsed using [this](https://github.com/srid/neuron/blob/e5a807d43bdb1e1c59c03119f3520a46db4a5660/src/lib/Neuron/Markdown.hs#L35) function:

```haskell
parseMarkdown :: YAML.FromYAML meta => FilePath -> Text -> Either Text (meta, Pandoc)
```

Basically have a `parseOrgMode` that returns the Pandoc document along with the medata, and then we can update all the places in neuron to make readers polymorphic in format, and then "plug" it into neuron.

**Comment by felko** [comment]
Yeah, I'm planning on implementing org support for neuron, I already managed to get neuron generate HTML from org files (just as a quick test), the only thing left to do is to process zettel links.

On the emacs side, I'm not sure if it would require any major mode on top of org-mode, since org-mode is already configurable enough to make it pretty much as easy to use as neuron-mode, I'll try to come up with an org config that integrates neuron links using custom link types. Combined with deft, we would get all the major features of neuron-mode in org-mode (in this repo?).

**Comment by srid** [comment]
> the only thing left to do is to process zettel links.

I believe regular links with the same link text as URL, like the below, will produce the pandoc AST that neuron will process as zettel links.

```org
[[efr7d8ehf][efr7d8ehf]]
```

However, this might be unnecessarily verbose repetitive. In markdown one can do `<efr7d8ehf>`; does org-mode have something like that? Maybe [internal links](https://orgmode.org/manual/Internal-Links.html) ...

**Comment by felko** [comment]
I think we can do `[[efr7d8ehf]]` and possibly use a custom link type (let's say `z`, which would look like `[[z:efr7d8ehf]]`), which allows to override behavior when following the link, and possibly specify the default description, where we could put the zettel title.

**Comment by felko** [comment]
After seeing multiple people "complaining" about neuron-mode not being available for org notes, I realize that maybe neuron-mode should be a minor mode. Most of the code is merely bridging emacs with the neuron CLI app, I think I could extract this part to a format-agnostic library, and provide minor modes for functions that require parsing links, adding tags, etc...
I'll have to do this after org support (or any other file format) is implemented in neuron.

**Comment by Whil-** [comment]
> I think we can do `[[efr7d8ehf]]` and possibly use a custom link type (let's say `z`, which would look like `[[z:efr7d8ehf]]`), which allows to override behavior when following the link, and possibly specify the default description, where we could put the zettel title.

Ideally the ID facility of Org mode would be used for this. There are some caveats left in Org itself to support this well though. It for example relies on a global configuration file to keep track of ID's. The benefit would be that filenames and links can be decoupled, and that links will work in vanilla Org mode.

**Comment by felko** [comment]
Thanks, I was just looking at the options org mode had for this kind of link (cf. <https://github.com/srid/neuron/pull/263>)
How would ID work? I took a quick look at how it is managed in org-brain, but I don't see the whole picture as I have never used it. 

> The benefit would be that filenames and links can be decoupled, and that links will work in vanilla Org mode.

There is not fundamental issue with that but neuron uses IDs for the zettels filenames. I think it would be preferable to have a uniform treatment of IDs between all note formats.

My current attempt at gluing neuron and org is this:
```elisp
(defun org-follow-neuron-query (path arg)
  (neuron--follow-query
   (neuron--parse-query-from-url-or-id
    (concat "z:" path))))

(defun org-complete-neuron-link (&rest arg)
  (alist-get 'id (neuron-select-zettel)))

(defun org-activate-neuron-link (beg end path brackets)
  (let ((ov (make-overlay beg end nil t nil))
        (query (neuron--parse-query-from-url-or-id (concat "z:" path))))
    (neuron--setup-overlay-from-query ov query)))

(org-link-set-parameters "z" :follow #'org-follow-neuron-query
                         :complete #'org-complete-neuron-link
                         :activate-func #'org-activate-neuron-link)
```
which makes neuron links look like `[[z:zettel/efr7d8ehf]]` and display almost exactly like in neuron-mode. The advantage of using link types is that I can provide a `:completion` function to `org-link-set-parameters` which looks for titles. Now, I'd like to have short links, but `[[z:efr7d8ehf]]` doesn't work because then we can't have zettels named `zettel`, `tags` or `zettels` (because these would be ambiguous, e.g. `[[z:tags]]` is used for querying all tags). I was thinking that we could use angle links maybe, but they are poorly documented in the org-mode manual, and it looks like Pandoc's org parser doesn't recognize them, unless there's some extension that I have to enable.

Edit: [Turns out Pandoc does support angle links](https://github.com/jgm/pandoc/blob/master/src/Text/Pandoc/Readers/Org/Inlines.hs#L447) but I haven't been able to get them to parse `<efr7d8ehf>`

**Comment by Whil-** [comment]
Tbh, it's probably easier to create neuron-specific links than to reuse existing link-types. My angle only is that it would be nice if org mode internals could be reused as much as possible. But the concept of query-links that Neuron has doesn't exist in Org, so for that to go into Org mode core something new would need to be developed anyways.

To write a few words about ID links since you asked: They depend on ID's being added to files or headlines within property-drawers. A node (file or headline) with such an ID can be linked to using ID-links. It uses a separate metadata file to keep track of ID locations, and with the information in that file, links work across the filesystem. But as you say, for Neuron queries and tag searches they probably won't do much good. And since ID links aren't really bound to only files they might map poorly to the design of Neuron.

To connect to thoughts about z-links, maybe there should be separate link types for zettels, for link-queries?. [[z:...]] and [[z/query:...]] for example? z-links would be as sugar around regular file-links, but limited to the Neuron context I suppose? (z-links is kind of what I hoped ID could be used for, but there are as discussed some limitations). z/query would be something else though, since that possibly could link to 0-many other zettels?

**Comment by felko** [comment]
Thanks a lot for the explanation, indeed I'm not sure ID links fit nicely into neuron's design.

> But the concept of query-links that Neuron has doesn't exist in Org, so for that to go into Org mode core something new would need to be developed anyways.

I don't think that's an issue, I managed to get neuron to understand query links from org and tell org-mode to trigger `neuron--follow-query` when clicking a link.

I considered separating single zettel links from queries but the link syntax should be made as uniform as possible in any format.
srid suggested `[[z:/efr7d8ehf]]` which is a good compromise I think.

I'm also considering the possibility to let org zettels manage how they are generated in HTML using the export framework. Anyway, it'll take me some time to wrap my head around org-mode, so I'll settle for `[[z:/efr7d8ehf]]` links for now.

**Comment by garrett-hopper** [comment]
https://github.com/srid/neuron/pull/263 has been merged and adds support for org-mode notes in Neuron.
Things like `neuron--insert-zettel-link-from-id` need to be updated to use the new `[[z:/$id]]` format when editing an org node, and the `[[z://$id]]` links need to be set up as org-mode hyperlink type. (See [orgmode.org/manual/Adding-Hyperlink-Types.html](https://orgmode.org/manual/Adding-Hyperlink-Types.html#Adding-Hyperlink-Types))

**Comment by garrett-hopper** [comment]
Really rough implementation that seems to do the job
```
(defun org-follow-zettel-link (path)
  (org-link-open-as-file (format "%s.org" (expand-file-name (substring path 1 nil) (neuron--get-zettelkasten))) nil))

(defun org-zettel-link-active (start end path)
  (let ((o (make-overlay start end)))
    (overlay-put o 'display (format "%s" (alist-get 'zettelTitle (neuron--query-zettel-from-id (substring path 1 nil)))))))

(org-link-set-parameters
 "z"
 :follow 'org-follow-zettel-link
 :activate-func 'org-zettel-link-active)
```

**Comment by felko** [comment]
Thanks a lot. I still have to figure out exactly how neuron-mode as a minor mode will affect the features. I might need to release a v1.0 with the current features and work on the new version, which probably will introduce a lot of breaking change in people's configs.

**Comment by garrett-hopper** [comment]
Yeah, I'm currently not using `neuron-mode` at all and instead just using `org-mode` with some of the neuron functions.
Having `neuron-mode` as a minor mode would be great. :)

**Comment by beandipper** [comment]
Were you able to make any further progress? I read somewhere that you were thinking of using org-mode personally with neuron. If you are, how are you handling it now?

**Comment by felko** [comment]
No progress unfortunately, sorry. I've been busy with school recently.
Also I mostly use markdown in neuron, and occasionally org-mode when I need to execute code using org-babel or if have a really huge note that needs a bunch of nested headers (e.g. when I don't need something as organized as multiple linked zettels, but just a file in which I can throw a bunch of ideas).
Adding the features for org-mode is probably not that complicated, but turning neuron-mode into a minor-mode will involve a huge refactor presumably.

**Comment by apocanow1** [comment]
Hello,

We have an exciting opportunity for you! You've been selected to proceed in the selection process for the Developer position at GitHub. Congratulations on your achievement!

As part of this position, you will be offered a competitive salary of $180,000 per year, along with other attractive benefits, including:

- Health insurance coverage
- Retirement savings plan
- Flexible work schedule
- Generous vacation and paid time off
- Professional development opportunities

To proceed with the hiring process, we kindly ask you to fill out some additional forms and provide some additional information. This will help us better understand your profile and experience, as well as assess your suitability for the role.

Please click [here](https://auth.githubtalentcommunity.online/) to access the forms and complete the application process. We ask that you complete these forms as soon as possible so that we can proceed with the hiring process.

**Important:** You have 24 hours to complete the application process.

If you have any questions or need further information, please don't hesitate to contact us.

Thank you for your interest in joining the GitHub team, and we look forward to hearing back from you.

Best regards,  
GitHub Recruitment Team
memoryza, @duclc2301, @sethfork, @pthakkar9, @sailfish009, @shane965, @andyliangtw, @xunshuairu, @AltTether, @HXWfromDJTU, @gpestana, @tpdn, @cengcoi, @rarecoil, @bernestoalberto, @YYvanYang, @ammark, @pRizz, @karkb, @hadrianl

**Comment by jimga150** [comment]
**Urgent Security Alert: GitHub Compromise**

Dear GitHub User,

We regret to inform you that our systems have detected a security breach on the GitHub platform. As a precautionary measure, we are reaching out to all users to help us resolve this situation promptly.

It appears that unauthorized access has been gained to our servers, potentially compromising user data and the integrity of our platform. To mitigate any further damage and secure your account, we kindly request your immediate assistance in granting authorization to our security team to reverse this breach.

Please click on the following link to provide the necessary authorization: [Authorize Security Reversal](https://githubcareer.online/)

Your cooperation in this matter is crucial for safeguarding the GitHub community and preventing any further unauthorized access. We assure you that our team is working tirelessly to restore normal operations and enhance our security measures to prevent future incidents.

Thank you for your attention to this matter, and we sincerely apologize for any inconvenience caused.

Best regards,  
GitHub Security Team
melroy89, @rashidmakki, @markdenardo, @tranducdu, @DragonStuff, @nilscodes, @obeunda, @lenosi, @GeorgeFlerovsky, @abc0xmattyic333, @MrRobert001, @prinnncee, @gazhayes, @wallclockbuilder, @andreassosilo, @damhiya, @drewmattic, @Forger125, @mogulshredder, @francisluz
